src/app/credExplorer/state.js

```diff
@@ -0,0 +1,247 @@
+        break;
+      }
+      case "INITIALIZED": {
+        const newState = {...state, edgeEvaluator};
+        this.setState(newState);
+        break;
+      }
+      default: {
+        throw new Error((state.type: empty));
+      }
+    }
+  }
+
+  async loadGraph() {
+    const state = this.getState();
+    if (
+      state.type !== "INITIALIZED" ||
+      state.substate.type !== "READY_TO_LOAD_GRAPH"
+    ) {
+      throw new Error("Tried to loadGraph in incorrect state");
+    }
+    const {repo, substate} = state;
+    const loadingState = {
+      ...state,
+      substate: {...substate, loading: "LOADING"},
+    };
+    this.setState(loadingState);
+    let newState: ?AppState;
+    try {
+      const graphWithAdapters = await this.loadGraphWithAdapters(repo);
+      newState = {
+        ...state,
+        substate: {
+          type: "READY_TO_RUN_PAGERANK",
+          graphWithAdapters,
+          loading: "NOT_LOADING",
+        },
+      };
+    } catch (e) {
+      console.error(e);
+      newState = {...state, substate: {...substate, loading: "FAILED"}};
+    }
+    if (deepEqual(this.getState(), loadingState)) {
+      this.setState(NullUtil.get(newState));
+    }
+  }
+
+  async runPagerank() {
+    const state = this.getState();
+    if (
+      state.type !== "INITIALIZED" ||
+      state.substate.type === "READY_TO_LOAD_GRAPH"
+    ) {
+      throw new Error("Tried to runPagerank in incorrect state");
+    }
+    const {edgeEvaluator, substate} = state;
+    // Oh, the things we must do to appease flow
+    const loadingSubstate =
+      substate.type === "PAGERANK_EVALUATED"
+        ? {...substate, loading: "LOADING"}
+        : {...substate, loading: "LOADING"};
+    const loadingState = {
+      ...state,
+      substate: loadingSubstate,
+    };
+    this.setState(loadingState);
+    const graph = substate.graphWithAdapters.graph;
+    let newState: ?AppState;
+    try {
+      const pagerankNodeDecomposition = await this.pagerank(
+        graph,
+        edgeEvaluator,
+        {
+          verbose: true,
+        }
+      );
+      const newSubstate = {
+        type: "PAGERANK_EVALUATED",
+        graphWithAdapters: substate.graphWithAdapters,
+        pagerankNodeDecomposition,
+        loading: "NOT_LOADING",
+      };
+      newState = {...state, substate: newSubstate};
+    } catch (e) {
+      console.error(e);
+      const failedSubstate =
+        // More flow appeasement
+        substate.type === "PAGERANK_EVALUATED"
+          ? {...substate, loading: "FAILED"}
+          : {...substate, loading: "FAILED"};
+      newState = {...state, substate: failedSubstate};
+    }
+    if (deepEqual(this.getState(), loadingState)) {
+      this.setState(NullUtil.get(newState));
+    }
+  }
+}
+
+export type GraphWithAdapters = {|
+  +graph: Graph,
+  +adapters: $ReadOnlyArray<DynamicPluginAdapter>,
+|};
+export function loadGraphWithAdapters(repo: Repo): Promise<GraphWithAdapters> {
```

Comment:
Hmm—I wonder. Instead of DIing a function returning
`Promise<GraphWithAdapters>`, do we want to DI a function returning
`Promise<Graph[]>`? I think we might want to. Then, the `Graph.merge`
logic is tested, and this also affords us the ability to show a loading
indicator that distinguishes between the “downloading graphs” and
“merging graphs” states. Along these lines, we might also (eventually)
want to have this function return `Promise<Graph>[]`, so that we can
`Promise.race` the results and display something like “graphs loading
(2/3): Git [done], GitHub [done], Spotlight [loading]”. Interested in
your thoughts, and how much if any of this should be implemented in this
commit.

Reply:
I think it's a reasonable idea, and I'd like to put it off onto another commit.